### DISCUSSION TITLE
Python 3.11 venv altinstall in EV3dev Debian 10 'Buster' experimental image

### DISCUSSION THREAD

**Discussion by Growflavor** [ev3dev/ev3dev#1634] [Ideas]
The EV3 does benefit dramatically from the performance improvements in Python 3.11:

```
Loops per second [EV3fast](https://github.com/QuirkyCort/ev3dev-lang-python-fast) Buster python 3.7:

Starting: brickrun -r --directory="/home/robot/Active" "/home/robot/Active/EV3devMixTest.py"
----------
236.40667471456294

Loops per second EV3fast Buster python 3.11.9:

Starting: brickrun -r --directory="/home/robot/Active" "/home/robot/Active/EV3devMixTestPy11.py"
----------
Import warning: Failed to import "evdev". Button will be unusable!
341.9160269254817
```

![Python-3 11-Performance-Benchmarks--2-](https://github.com/ev3dev/ev3dev/assets/1580559/ca854b56-254f-4f09-9d4d-511c08e64990)
source: https://flyaps.com/blog/python-3-11-release-features-overview/

[Python 3.11 vs Python 3.12 – performance testing](https://en.lewoniewski.info/2023/python-3-11-vs-python-3-12-performance-testing/)
![EV3devBuster](https://github.com/ev3dev/ev3dev/assets/1580559/0234cd34-969c-40f0-a0ea-ff62045402fa)

A **Python 3.11 venv** is included in the following Debian 10 "Buster" image that brings together into one image dlech's prior work.

This is currently the 'goto' EV3dev image that my son & I use for all our EV3 projects...

Try out your favorite programs & run test programs to evaluate the 'if useable' functionality of features introduced since python 3.5 in the Python 3.11 venv ... all that is needed is to change the shebang at the start of your python file to:

**#!/home/robot/venv311/bin/python3.11**

**Please reply to this discussion with the results (& link to your code if available online)!**

A) Image to flash to SDcard via Balena Etcher available for download:

NOTE: [please see the comment below for details on the reboot during just the FIRST Boot after flash](https://github.com/orgs/ev3dev/discussions/1634#discussioncomment-9890535).

https://drive.google.com/file/d/1Y8g46pREtF-7YdmpBeYbCxqt2HyAJkv0/view?usp=drive_link
     ev3dev10img02b.zip SHA-256: B9E7D392579570023441A20236C0E2398FFF3B7267C1C20C5CBE996D0FF692D3
     ev3dev10img02b.img SHA-256: B33787680AC4575D59A7C739EB531E4F4F553AF230B33F25A87D84A1FD5DAD72

B) here is a link to the docker image used to make the above image so that you can modify packages to meet your needs & build your own sdcard images using Brickstrap for this Debian 10 "Buster" draft v02:

    docker pull growflavor/ev3images:ev3dev10imgv02b

    sha256:59416d96e7f9ad30b5da16a6c9e6b09d2e71e404880202965c23cd5604d74ac7

--the py3.11 venv pip builds wheels from source...so use docker to add more packages to the venv...

--using apt is painful onboard the EV3, so easier to: 

<details>

  <summary>update the image using a docker container:</summary>

--make changes/updates in a docker container on a machine with a favorite brickstrap/docker setup:

commands below are for docker running on linux...
& in this case sudo is used since it is easier with Ubuntu running via Windows WSL2

details about brickstrap:  https://github.com/ev3dev/brickstrap

  ```javascript
sudo docker pull growflavor/ev3images:ev3dev10imgv02
sudo docker tag growflavor/ev3images:ev3dev10imgv02 ev3d10py311old
sudo docker run --platform linux/arm/v5 -it ev3d10py311old su -l robot
  ```   
--you will get a robot@ prompt, make the updates/changes you want inside this container, then logout from the robot@ prompt
 
--then commit the docker container you just updated to a docker image (docker ps -a will show the image(s) & ID#):

  ```javascript
      sudo docker ps -a
      sudo docker container commit ID#_of_your_updated_container_e.g.123eeed582c2 ev3d10py311new:latest
  ```  
--then use [brickstrap](https://github.com/ev3dev/brickstrap) to make a new .img from your updated docker container to flash with Etcher, for example:

  ```javascript
sudo brickstrap create-tar ev3d10new:latest ev3dev10py311new.tar

sudo brickstrap create-image ev3dev10py311new.tar ev3dev10py311new.img
  ```  

You may return to your stopped container (use your container's name i.e. not quizzical_wright ) to make further changes/updates anytime by starting it:

  ```javascript
sudo docker start -i quizzical_wright
  ```  
or if it is still running you can activate a shell to resume work with syntax like:

  ```javascript
sudo docker exec -it quizzical_wright /bin/bash
  ```  
you can also copy files out of the container with syntax like this:

  ```javascript
sudo docker cp quizzical_wright:/usr/local/src/brickman_0.11.0+fl0wey1_armel.deb .
  ```  
</details>

**Some details about the "Buster" experimental image:**

robot@ev3dev:$ python --version
Python 2.7.16
robot@ev3dev:$ python3 --version
Python 3.7.3
robot@ev3dev:$ python3.11 --version
Python 3.11.9

robot@ev3dev:$ source venv311/bin/activate
(venv311) robot@ev3dev:$ python --version 
Python 3.11.9
(venv311) robot@ev3dev:$ 

(venv311) robot@1d282c75b987:~$ pip list
Package            Version
------------------ -----------
beniget            0.4.1
certifi            2024.6.2
cmake              3.28.3
compilers          0.0.0
contourpy          1.2.1
cppy               1.2.1
cycler             0.12.1
Cython             3.0.10
fonttools          4.53.0
gast               0.5.4
kiwisolver         1.4.5
matplotlib         3.8.4
meson              1.4.1
meson-python       0.16.0
ninja              1.10.2.3
numpy              1.25.2
packaging          24.1
pillow             10.3.0
pip                24.0
plumbum            1.8.3
ply                3.11
pybind11           2.12.0
pyblas             0.0.10
pyparsing          3.1.2
pyproject-metadata 0.8.0
python-dateutil    2.9.0.post0
python-ev3dev      1.2.0
python-ev3dev2     2.1.0.post1
pythran            0.16.1
rpyc               5.3.1
setuptools         70.0.0
setuptools-scm     8.1.0
six                1.16.0
wheel              0.43.0
(venv311) robot@1d282c75b987:~$

**To use the python 3.11 venv just change the shebang at the start of your python file to:**

#!/home/robot/venv311/bin/python3.11

**You may also RPyC to the venv:**

venv RPyC server bluetooth example:
python3.11 ./py11venv/bin/rpyc_classic.py --host=169.254.109.129

venv RPyC server LAN example:
python3.11 ./py11venv/bin/rpyc_classic.py --host=192.168.0.103

 -- VScode has a jupyter note book extension so you can have a notebook open directly in VScode served from the [same .venv you use for the pylance interpreter](https://github.com/ev3dev/vscode-hello-python?tab=readme-ov-file#code-completion) & [use RPyC 5.3.1 to directly control your EV3 by notebook code cells](https://i.imgur.com/PWo6Qx2.png) 

you may also RPyC to the 'Buster' system Python 3.7:

robot@ev3dev:~$ hostname -I
192.168.1.110 

then activate the EV3 RPyC server using the IP shown by hostname -I: 
sudo python3 /usr/local/bin/rpyc_classic.py --host=192.168.1.110

<details>
  <summary>& of course, some common 'sudo apt install' Python 3.7 system packages as well in this image:</summary>
  
pybricks-micropython: GyroBoy runs OK 
openrobertalab: connects
& Just using pip to list...these are **system package manager 'apt' installed** (except for RPyC==5.3.1)
  
  ```javascript
robot@ev3dev:~$ python3 -m pip list
Package         Version
--------------- -----------
agt             1.0
asn1crypto      0.24.0
atomicwrites    1.1.5
attrs           18.2.0
beautifulsoup4  4.7.1
chardet         3.0.4
cryptography    2.6.1
cycler          0.10.0
decorator       4.3.0
entrypoints     0.3
evdev           1.1.2
html5lib        1.0.1
joblib          0.13.0
keyring         17.1.1
keyrings.alt    3.1.1
kiwisolver      1.0.1
lxml            4.3.2
matplotlib      3.0.2
more-itertools  4.2.0
nose            1.3.7
numexpr         2.6.9
numpy           1.16.2
openrobertalab  1.7.4
pandas          0.23.3+dfsg
Pillow          5.4.1
pip             24.0
pixy            0.3
pluggy          0.8.0
plumbum         1.8.3
protobuf        3.6.1
psutil          5.5.1
ptvsd           5.0.0a5
py              1.7.0
PyBluez         0.22
pycrypto        2.6.1
PyGObject       3.30.4
pyparsing       2.2.0
pyserial        3.4
pytest          3.10.1
python-dateutil 2.7.3
python-ev3dev   1.2.0
python-ev3dev2  2.1.0
pytz            2019.1
pyudev          0.21.0
pyxdg           0.25
rpyc            5.3.1
scikit-learn    0.20.2
scipy           1.1.0
SecretStorage   2.3.1
setuptools      40.8.0
simplejson      3.16.0
six             1.12.0
soupsieve       1.8
tables          3.4.4
webencodings    0.5.1
wheel           0.32.3
robot@ev3dev:~$ 

  ```
  </details>

Some additional notes:

--the above image is a make of the default Py3.11 ./configure (& only module _dbm did not build) 
    so if you want to make & altinstall a different Py3.11 ./configure such as minimal
        https://docs.python.org/3.11/using/configure.html
    then look at /usr/local/src in this image:

docker pull growflavor/ev3images:ev3dev10makepy11

--If you encounter memory limits, then I suggest stopping the openrobertalab service when you want to use the venv:
sudo systemctl stop openrobertalab

**Comment by Nikolay-Zagrebin** [comment]


**Reply by Growflavor** [reply]
to shorten & make your posts easier to see, would you please edit to delete the "quote" part (or only keep a few key few lines of the quote)?  Thanks!

**Reply by Nikolay-Zagrebin** [reply]
> to shorten & make your posts easier to see, would you please edit to delete the "quote" part (or only keep a few key few lines of the quote)? Thanks!

In principle, you have already answered me, you can already delete it here

**Comment by Growflavor** [comment]
1) RE: "reply to this discussion with the results (& link to your code if available online)!"

Here is a list of simple files we have run so far to do a basic check the ev3dev packages in the 'Buster' Python 3.11 venv:

https://github.com/Growflavor/EV3projects/tree/main/Python311_venv_Tests

GyroBoy on EV3dev
ColorSensorTest
EV3devFastMixTest -- **275 loops per second :-)  so less than 4ms per loop**
EV3devFastTest
EV3voltageCurrentTest
EV3speakTest
GyroDriveTankMotorTest
numpyTest

So far, the only issue we have encountered remains the message about buttons:

```
'Starting: brickrun -r -directory="/home/robot/Active" "/home/robot/Active/EV3devMixTestPy11.py"
'----------
Import warning: Failed to import "evdev". Button will be unusable!
275.81100512581344  loops per second :-)
'>>> 
sometimes even:
341.9160269254817  loops per second!  Have not figured out why varies from 268 to 341...
'>>> 
```

2)
@Nikolay-Zagrebin 

RE: "EV3 itself has an interface with unreadable text ((("  

I speculate that one of the font or other packages has not been added or updated in your image...I had to add MANY packages to the base ev3dev Bullseye image...better if we follow up on that topic over at #1554 
https://github.com/ev3dev/ev3dev/issues/1554#issuecomment-2019205120

**Comment by Madelyn-of-Hell** [comment]
hi! I just tried to install this ev3dev image onto my robot, and ran into a confusing issue
when I tried to start the robot, it when part of the way through the start up sequence, then gave the message 'Press enter to continue.' None of the robot's buttons dismissed this, and i was unable to continue or turn the robot off (I ended up having to remove the power pack to turn it off).

would you be able to explain how to get around this issue?

**Reply by Growflavor** [reply]
Yes, the flash images built from the docker containers using Brickstrap do this only on the _FIRST boot_ after the SDcard flash.  
I think it is related to the OS configuring itself & expanding the file system to use the whole SDcard.

All future boots with the SDcard are normal.

On some builds, boot continues after just pressing the EV3 middle button = 'enter'.  However, with this "Buster" image, you do indeed need to force a reboot when it pauses at this point during the FIRST boot only: 

You do not need to remove the battery (although that also works ;-).

During future "first boot after flash", just press and hold down together the EV3 "back" & "middle" buttons until the screen turns off or LEDs change...then release the two buttons...the system will reboot.

![RebootEV3](https://github.com/ev3dev/ev3dev/assets/1580559/a91353e7-9a80-448c-8678-dc35c1d20307)

**Reply by Madelyn-of-Hell** [reply]
oh my goodness thank you so much!
I can't believe I didn't think to boot it again. Clearly, this project is wearing away at me.

**Comment by Growflavor** [comment]
### W.i.P 2024: Tracking Kernel Buildscripts Updates from Forks

https://github.com/ev3dev/ev3dev-buildscripts

**original kernel configs:** 
https://github.com/Growflavor/EV3projects/tree/main/EV3KernelConfigs

**kernel changes of interest:**

- kernels since 4.14 should support # CONFIG_ZRAM_WRITEBACK so that swapfile will be used if ZRAM gets full


**universal changes:**

- update git:// links to https://

** @StepanTheFlowey Buildscripts & Kernel Forks:**
https://github.com/StepanTheFlowey/ev3dev-buildscripts

- change in compiler

**Comment by gregorianrants** [comment]
when i follow the provided link to the image
https://drive.google.com/file/d/1In-FHMT9mu4jAle2686JfvUvR_KmldlV/view?usp=drive_link
ev3dev10img02b.zip SHA-256: B9E7D392579570023441A20236C0E2398FFF3B7267C1C20C5CBE996D0FF692D3
ev3dev10img02b.img SHA-256: B33787680AC4575D59A7C739EB531E4F4F553AF230B33F25A87D84A1FD5DAD72

I get the message:
Sorry, the file you have requested does not exist.

I am finding it impossible to install python packages in stretch as the debian packages do not seem to be available. specifically i am trying to install pyzmq which are python binding for the c package 0mq which i also cant install.

i want to try a newer release of debian to see if i have more luck.  there seems to be packages available so should be easier.  im a bit of a novice though for anything beyond pip install and using pip doesnt seem to be an option.  installing debian packages directly is a bit new to me and i am a bit unsure of what i am doing tbh.

**Reply by Growflavor** [reply]
Sorry about that, I have put that 02b version back & thus it has a new link but SHA-256 should be the same:

**ev3dev10img02b.zip**
https://drive.google.com/file/d/1Y8g46pREtF-7YdmpBeYbCxqt2HyAJkv0/view?usp=drive_link
__ev3dev10img02b.zip SHA-256__: B9E7D392579570023441A20236C0E2398FFF3B7267C1C20C5CBE996D0FF692D3
__ev3dev10img02b.img SHA-256__: B33787680AC4575D59A7C739EB531E4F4F553AF230B33F25A87D84A1FD5DAD72

I had accidentally removed the 02b version since I had been tinkering with version 02f  that has more venv premade while in docker 
(v02f is likely too bloated for your interest, but since flashing multiple EV3dev versions on different SDcards is easy, below is the link so anyone can try it out...):

**ev3dev10v02fclean.zip** 
https://drive.google.com/file/d/1CTgfor0x5ChCO9ZSYRA7_vibquguLeqN/view?usp=drive_link

note: "clean" means I had to delete the PIP cache so that Brickstrap made an image that fit within the ~4GB file system of the default EV3dev image & flashed OK with Balena Etcher...

it is possible to extend the file system manually, such as described at these links:

https://serverfault.com/questions/994448/extend-linux-partition-size-without-losing-any-data-techniques
https://unix.stackexchange.com/questions/67095/how-to-expand-ext4-partition-size-using-command-line
https://www.privex.io/articles/how-to-resize-partition/

or, in my case I make most of my backup images of my EV3dev SDcards using the commercial backup software Acronis True Image, which automatically resizes the image file system to the size of my SDcard when I 'restore' an Acronis image of an EV3dev

Unfortunately since I posted those versions I have been diagnosed with a problematic stage 4 cancer so my work with my son on the EV3 has been paused for the moment since every decision I make in the next 2 months will influence how long I live...I do hope to enjoy working with my son on the EV3 & be more active here again once I have worked out my medical path...

**Reply by gregorianrants** [reply]
Thank You, that is very kind of you to answer with everything you have going on.  your answer is most helpfull for what i am trying to do.

I wish you the absolute best, and hope things work out for you.  I would like nothing more than for you and your son to have some quality time enjoying your shared interest.
